[Guide] Rename default branch (#1676)
platformsh/platformsh-docs: docs/src/guides/general/default-branch.md

---

{{< description >}}

Soon you will be able to select your default branch during the project creation steps, but until that time this guide will show you how to manually make the switch yourself. You can complete some of these steps through the management console, but since all can be completed with the CLI those commands alone are listed. Be sure to [install the CLI](http://localhost:1313/development/cli.html#installation) if you have not already done so. It's assumed you are changing the default branch of your project on Platform.sh from `master` to `main`. If using another name for the default branch, update the commands accordingly. 

## New projects

The steps below assume:

- You have an external repository on GitHub, GitLab, or BitBucket you are trying to integrate with Platform.sh.
- That repository's default branch is not `master`.

### 1. Create a new project

First, create an empty project with the Platform.sh CLI command `platform create`:

```bash
$ platform create --title='Main Project' --region=us-3.platform.sh --plan=development --environments=3 --storage=5 --no-set-remote
```

> This will create a new project with Master as the parent environment by default. Modify the flags to fit your use case, otherwise exclude and the CLI will ask you to set them individually during creation. Copy the `Project ID` provided when the command completes, and substitute the value in the steps below. You can also visit the provided management console URL for the project (i.e. `https://console.platform.sh/<USER>/<Project ID>`) to verify the steps as you go along.

Reviewer: ```suggestion
This will create a new project with Master as the parent environment by default. Modify the flags to fit your use case, or skip them and the CLI will ask you to set them individually during creation. Copy the `Project ID` provided when the command completes, and substitute the value in the steps below. You can also visit the provided management console URL for the project (i.e. `https://console.platform.sh/<USER>/<Project ID>`) to verify the steps as you go along.
```

The project you just created is empty, that is, there is no code initialized on it's Master environment. You will need to have something on that environment to begin with in order to create the `main` branch, so you can initialize it with a template for now:

```bash
$ platform environment:init -p <Project ID> -e master https://github.com/platformsh-templates/hugo.git
```

Then, create the "Main" environment:

```bash
$ platform environment:branch main --title=Main -p <Project ID> --force
```

{{< note >}}
The CLI assume that you are running this command within a local copy of your repository, so the `--force` flag is included above to bypass that.
{{< /note >}}

### 2. Deactivate the Master environment

> Since Master is currently the parent environment, it is protected, and so the normal `platform environment:delete` command will not work at this stage. To get around this, place an authenticated cURL request on the project's API with the CLI command to deactivate it:

Reviewer: ```suggestion
Since Master is currently the root environment, it is protected, and so the normal `platform environment:delete` command will not work at this stage. To get around this, place an authenticated cURL request on the project's API with the CLI command to deactivate it:
```

```bash

> $ platform project:curl -X POST -p <Project ID> environments/master/deactivate 

Reviewer: We don't document or suggest using `platform:curl` anywhere else at this point.  It's a hack. Is that really as much as we're going to do?

Author: For now is my understanding, otherwise they'd have to place the API calls themselves and this is at least cleaner.

Reviewer 2: What about

```sh
platform environment:delete master
```

Author: @pjcdawkins I still get this message trying to deactivate master with this command

```
The main environment cannot be deleted.
```

Reviewer 2: Ah that's a bug. The `is_main` property is incorrectly cached on the environment. Run `platform project:curl system/restart -X POST` and you'll see the cache being cleared... and then this CLI command would work as intended.

Obviously that wouldn't be so helpful in documentation but.. do we document broken or work-in-progress things? :man_shrugging:

Author: @pjcdawkins So this seems a little inconsistent actually for me, even with the `system/restart`. It worked fine on an eu-3 project, but doesn't work at all on us-3. Any idea when the bug fix will be applied?

Otherwise. Maybe it's good enough for now to keep the `platform project:curl` commands here for now and replace when the fix is in?

cc @Crell

Author: ```
$ platform project:curl system/restart -X POST -p fw36zqmvvzzls
{
  "status": "OK", 
  "code": 200, 
  "_embedded": {
    "activities": []
  }
$ platform  environment:delete master --no-delete-branch -p fw36zqmvvzzls
The main environment cannot be deleted.
```

Reviewer: As long as we know that a better alternative is coming, I'm OK with showing `project:curl` for now, provided it's marked as non-standard.  But that should be seen as extra incentive to finish the better alternative, not an excuse not to. :smile:

```

### 3. Make "Main" the parent environment

First, update the project's `default_branch` property with another authenticated request:

```bash

> $ platform project:curl -X PATCH -p <Project ID> -d '{"default_branch": "main"}'

Reviewer 2: ```
platform project:info default_branch main
```

```bash
$ platform environment:info -p <Project ID> -e master parent main
$ platform project:curl -X DEL -p <Project ID> environments/master
```

You will then be left with a single parent environment: Main. 

![Default environment main](/images/management-console/defaultbranch-main.png "0.75")

### 4. Setup the integration 

Place the command to integrate Platform.sh with your external [GitHub](/integrations/source/github.md), [GitLab](/integrations/source/gitlab.md), or [BitBucket](/integrations/source/bitbucket.md) repository. Select the settings you would like for the integration, but in most cases the default settings are best. The last stage will provide a warning: 

```bash
Warning: adding a 'github' integration will automatically synchronize code from the external Git repository.
This means it can overwrite all the code in your project.

Are you sure you want to continue? [y/N] y
```

> which will override the code on the "Main" environment - the template you started with - with the contents of the `main` branch of your external repository. You can not consult the rest of the [migration steps](/tutorials/migrating.md) to finish migrating your site to Platform.sh.

Reviewer: ```suggestion
which will override the code on the "Main" environment - the template you started with - with the contents of the `main` branch of your external repository. You can now consult the rest of the [migration steps](/tutorials/migrating.md) to finish migrating your site to Platform.sh.
```

First, close any open pull requests and resubmit them against `main`. Be sure to rebase your local brances with `git rebase --onto main master` if you still plan on continuing to work on them after the default branch switch. Once you resubmit them, they will appear under the "main" environment on Platform.sh. 

![Resubmitted PRs](/images/management-console/resubmit-prs.png "0.75")

### 3. Change the default branch on the external Git service

Platform.sh supports external Git integrations to a number of services, so follow the linked instructions below for changing the default branch to `main` for your provider:

- [GitHub](https://github.com/github/renaming)
- [GitLab](https://docs.gitlab.com/ee/user/project/repository/branches/#default-branch)
- [BitBucket](https://community.atlassian.com/t5/Bitbucket-questions/How-to-change-MAIN-branch-in-BitBucket/qaq-p/977418#:~:text=In%20Bitbucket%20Cloud%2C%20please%20go,repository%20details%20%3E%3E%20Main%20branch.)

### 4. Deactivate the Master environment

> Master is currently the parent environment, and you will need to deactivate in order to make `main` into the new parent. First, place an authenticated cURL request on the project's API with the CLI command to deactivate it:

Reviewer: ```suggestion
Master is currently the root environment, and you will need to deactivate in order to make `main` into the new parent. First, place an authenticated cURL request on the project's API with the CLI command to deactivate it:
```
